### PR TITLE
fixed babel configuration for #208

### LINKED ISSUE #208: babel-loader rules ignores additional_paths that come from node_modules
Ruby version: 3.1.2
Rails version: 7.0.4
Shakapacker version: 6.5.4

**Expected behavior:**
Directories in `webpacker.yml` `additional_paths` should all be transpiled, even if they're  inside `node_modules`. Otherwise third-party packages like let's say `@hotwired/turbo-rails` might not work for custom browserlist configurations. (I had to support Safari 14, and turbo needed to be transpiled because of this)

**Actual behavior:**
All directories inside `node_modules` are ignored by the babel-loader, even if they're set in `additional_paths`, like `node_modules/@hotwired/turbo-rails`

---

**Here's my modified rule, that fixes this (refactor it as you will, also the imports):**

The babel-loader rule includes all inclusions and excludes all node_modules by default, besides the ones that are in the inclusions.

```js
const { resolve } = require('path')
const { realpathSync } = require('fs')
const { loaderMatches } = require('shakapacker/package/utils/helpers')

const {
  source_path: sourcePath,
  additional_paths: additionalPaths,
  webpack_loader: webpackLoader,
} = require('shakapacker/package/config')
const { isProduction } = require('shakapacker/package/env')

const inclusions = [sourcePath, ...additionalPaths].map(p => {
  try {
    return realpathSync(p)
  } catch (e) {
    return resolve(p)
  }
})

/**
 * Modified babel-loader config of `node_modules/shakapacker/package/rules/babel.js`.
 * We had to modify the loader to NOT exclude inclusions (this is in a way a bug of shakapaker).
 * When we exclude inclusions (in webpacker.yml @ additional_paths) that might be inside node_modules,
 * like node_modules/@hotwired/turbo-rails, they won't be transpiled!
 *
 * See:
 * - https://webpack.js.org/configuration/module/#ruleexclude 
 * - https://github.com/babel/babel-loader/issues/171 
 * - Original rule: node_modules/shakapacker/package/rules/babel.js
 *
 * @type {*}
 */
module.exports = loaderMatches(webpackLoader, 'babel', () => ({
  test: /\.(js|jsx|mjs|ts|tsx|coffee)?(\.erb)?$/,
  include: inclusions,
  exclude: [
    {
      // exclude all node_modules from running through babel-loader
      and: [resolve('node_modules')],
      // Do not exclude inclusions, as otherwise these won't be transpiled
      not: [...inclusions],
    },
  ],
  use: [
    {
      loader: require.resolve('babel-loader'),
      options: {
        cacheDirectory: true,
        cacheCompression: isProduction,
        compact: isProduction,
      },
    },
  ],
}))
```

**Usage:**

You can then use `addional_paths` for `node_modules` as well, example:

```rb
# webpacker.yml
additonal_paths = [
  'app/assets', 
  'engine/foo/app/assets',
  ...
  'node_modules/@hotwired/turbo',
  'node_modules/@hotwired/turbo-rails',
  'node_modules/@hotwired/stimulus',
  'node_modules/stimulus-use',
  ...
]
```

### PR DESCRIPTION
**Pull request by vaukalak**
In specific scenarios, you would need to transpile some of `node_modules`. This addition allows to specify `node_modules` to include via `webpacker.yml => additional_paths` configuration.

closes #208

### CODE REVIEW

File: package/rules/__tests__/__utils__/webpack.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+module.exports = {
+    createTrackLoader,
+    node_modules,
+    node_modules_included,
+    app_javascript,
+    createInMemoryFs,
+    createTestCompiler,
+}
```
Comment (justin808) [inline package/rules/__tests__/__utils__/webpack.js:48]: editor setup is incorrect

needs new line

File: package/rules/__tests__/babel.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    test('exclude node_modules', async () => {
+        const ignored = `${node_modules}/a.js`
+        const [tracked, loader] = createTrackLoader();
+        const compiler = createTestCompiler(createWebpackConfig(ignored, loader));
+        await compiler.run();
+        expect(tracked[ignored]).toBeUndefined();
+    });
+
+    test('process included node_modules', async () => {
```
Comment (justin808) [inline package/rules/__tests__/babel.js:52]: we should never transpile all of node_modules. Are you checking for only the additional paths?

File: package/rules/__tests__/babel.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        const included = `${node_modules_included}/a.js`
+        const [tracked, loader] = createTrackLoader();
+        const compiler = createTestCompiler(createWebpackConfig(included, loader));
+        await compiler.run();
+        expect(tracked[included]).toBeTruthy();
```
Comment (justin808) [inline package/rules/__tests__/babel.js:57]: Need to check that that the files outside of node_modules/included and inside of node_modules are not transpiled, in at least one test.

File: package/rules/esbuild.js [hunk 4]
```diff
@@ -1,23 +1,12 @@
...
 const {
-  source_path: sourcePath,
-  additional_paths: additionalPaths,
   webpack_loader: webpackLoader
 } = require('../config')
+const jscommon = require('./jscommon')
 
 module.exports = loaderMatches(webpackLoader, 'esbuild', () => ({
   test: /\.(ts|tsx|js|jsx|mjs|coffee)?(\.erb)?$/,
-  include: [sourcePath, ...additionalPaths].map((p) => {
```
Comment (justin808) [inline package/rules/esbuild.js:14]: Nice to dry this up! 👏

**Review by justin808 — CHANGES_REQUESTED** [review]
1. Create a new setting, like `packages_to_transpile`
2. Don't overload `additional_paths` 
3. Needs documentation and changelog

File: package/rules/__tests__/babel.js (hunk 3 above)
Comment (vaukalak) [reply package/rules/__tests__/babel.js:57]: @justin808 we do have such test: https://github.com/shakacode/shakapacker/pull/240/files/d5b2d1744eaf17915aa36dd842e3d27ca414b317#diff-45330288787adaa7643887ecb63f58775bebe604da29c9789f0d5444a12a4ce7R44

File: package/rules/__tests__/babel.js (hunk 2 above)
Comment (vaukalak) [reply package/rules/__tests__/babel.js:52]: @justin808 sure we do, and also we do have a test case for this. I'll work on a better verbiage.

File: package/rules/__tests__/__utils__/webpack.js (hunk 1 above)
Comment (vaukalak) [reply package/rules/__tests__/__utils__/webpack.js]: @justin808 thanks fixed, and created a new issue: https://github.com/shakacode/shakapacker/issues/242

**Comment by vaukalak** [comment]
> 1. Create a new setting, like `packages_to_transpile`
> 2. Don't overload `additional_paths`
> 3. Needs documentation and changelog

@justin808 I've updated the changelog, but I'm not sure how adding additional property `packages_to_transpile` would be helpful. The code of webpack rules wouldn't change from what we have in this PR. We do override `additional_paths` only during tests. The only real difference would be in the configuration:

```yml
additional_paths: ['app/assets', 'node_modules/to_transpile']
```
vs
```yml
additional_paths: ['app/assets']
packages_to_transpile: ['node_modules/to_transpile']
```

So `packages_to_transpile` would become the same thing as `additional_paths`, however with `node_modules` support. Those properties will be merged anyway when creating the configuration for webpack.  So, I don't know how it can be helpful, IMHO it will only add more confusion.

**Comment by justin808** [comment]
Let's update the branch and merge it.

**Pull request merged**